### PR TITLE
Enable GCM and GCL instead of InfluxDB on GCE

### PR DESCRIPTION
**Pull request by saad-ali**
- Enabled Google Cloud Monitoring (GCM) and Google Cloud Logging (GCL) instead of InfluxDB for Google Compute Engine (GCE) deployments.
- Modified cluster addon setup process to take JSON files in addition to YAML
- Converted all Heapster Kubernetes config files to JSON
- Get rid of all Heapster v1beta1 configs

CC @vishh

### CODE REVIEW

**Comment by vishh** [comment]
cc @zmerlynn for salt config changes.

**Comment by saad-ali** [comment]
E2E is green:
```
Ran 38 of 43 Specs in 1820.331 seconds
SUCCESS! -- 38 Passed | 0 Failed | 1 Pending | 4 Skipped I0504 23:12:37.766365   23115 driver.go:96] All tests pass

```

File: cluster/addons/cluster-monitoring/google/heapster-service.json [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+      "name": "heapster",
+      "kubernetes.io/cluster-service": "true"
+    },
+    "name": "monitoring-heapster"
+  },
+  "spec": {
+    "ports": [
+      {
+        "port": 80,
+        "targetPort": 8082
```
Comment (a-robinson) [inline cluster/addons/cluster-monitoring/google/heapster-service.json]: Shouldn't the heapster rc expose this port?

File: cluster/gce/config-default.sh [hunk 2]
```diff
@@ -91,7 +91,10 @@ ENABLE_DOCKER_REGISTRY_CACHE=true
 ENABLE_NODE_MONITORING="${KUBE_ENABLE_NODE_MONITORING:-true}"
 
 # Optional: When set to true, heapster, Influxdb and Grafana will be setup as part of the cluster bring up.
-ENABLE_CLUSTER_MONITORING="${KUBE_ENABLE_CLUSTER_MONITORING:-true}"
+ENABLE_CLUSTER_MONITORING="${KUBE_ENABLE_CLUSTER_MONITORING:-false}"
+
+# Optional: When set to true, heapster, Google Cloud Metrics and Google Cloud Logging will be setup as part of the cluster bring up.
+ENABLE_GOOGLE_CLUSTER_MONITORING="${KUBE_ENABLE_GOOGLE_CLUSTER_MONITORING:-true}"
```
Comment (a-robinson) [inline cluster/gce/config-default.sh]: This may be better as MONITORING_DESTINATION to match the logging options

File: cluster/gce/config-default.sh [hunk 3]
```diff
@@ -91,7 +91,10 @@ ENABLE_DOCKER_REGISTRY_CACHE=true
 ENABLE_NODE_MONITORING="${KUBE_ENABLE_NODE_MONITORING:-true}"
 
 # Optional: When set to true, heapster, Influxdb and Grafana will be setup as part of the cluster bring up.
-ENABLE_CLUSTER_MONITORING="${KUBE_ENABLE_CLUSTER_MONITORING:-true}"
+ENABLE_CLUSTER_MONITORING="${KUBE_ENABLE_CLUSTER_MONITORING:-false}"
```
Comment (a-robinson) [inline cluster/gce/config-default.sh]: If this is false, does heapster still get brought up?

File: cluster/gce/config-test.sh [hunk 4]
```diff
@@ -50,9 +50,13 @@ ENABLE_DOCKER_REGISTRY_CACHE=true
 # Optional: Install node monitoring.
 ENABLE_NODE_MONITORING="${KUBE_ENABLE_NODE_MONITORING:-true}"
 
-# Optional: When set to true, heapster will be setup as part of the cluster bring up.
+# Optional: When set to true, heapster, Influxdb and Grafana will be setup as part of the cluster bring up.
 ENABLE_CLUSTER_MONITORING="${KUBE_ENABLE_CLUSTER_MONITORING:-true}"
 
+# Optional: When set to true, heapster, Google Cloud Metrics and Google Cloud Logging will be setup as part of the cluster bring up.
+ENABLE_GOOGLE_CLUSTER_MONITORING="${KUBE_ENABLE_GOOGLE_CLUSTER_MONITORING:-false}"
```
Comment (a-robinson) [inline cluster/gce/config-test.sh]: same comments as in config-default.sh

**Comment by a-robinson** [comment]
What's the motivation for switching to JSON? Just a few days ago @bgrant0607 wrote that "We should just use YAML for examples. It's generally more user-friendly than JSON, and supports comments."

File: cluster/addons/cluster-monitoring/influxdb/heapster-controller.json [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+      },
+      "spec": {
+        "containers": [
+          {
+            "image": "gcr.io/google_containers/heapster:v0.11.0",
+            "name": "heapster",
+            "env": [
+              {
+                "name": "FLAGS",
+                "value": "--poll_duration=2m --stats_resolution=1m --sink influxdb:http://$MONITORING_INFLUXDB_SERVICE_HOST:$MONITORING_INFLUXDB_PORT_8086_TCP_PORT"
```
Comment (vishh) [inline cluster/addons/cluster-monitoring/influxdb/heapster-controller.json]: Why use env variables? Why not use DNS? This could be `http://monitoring-influxdb:80`.
The advantage with dns is that all these services can be started in parallel.

File: cluster/gce/config-default.sh (hunk 3 above)
Comment (vishh) [inline cluster/gce/config-default.sh]: Why not choose between, `['none' | 'influxdb' | 'google']`? 
And +1 to @a-robinson's question.

File: cluster/addons/cluster-monitoring/google/heapster-service.json (hunk 1 above)
Comment (saad-ali) [inline cluster/addons/cluster-monitoring/google/heapster-service.json]: Good point. The heapster service is not needed, I will remove it from both influxdb and google.

File: cluster/gce/config-default.sh (hunk 2 above)
Comment (saad-ali) [inline cluster/gce/config-default.sh]: Good idea, done.

File: cluster/gce/config-default.sh (hunk 3 above)
Comment (saad-ali) [inline cluster/gce/config-default.sh]: Changed, to choice of target.

File: cluster/addons/cluster-monitoring/influxdb/heapster-controller.json (hunk 5 above)
Comment (saad-ali) [inline cluster/addons/cluster-monitoring/influxdb/heapster-controller.json]: I don't like the idea of setting the port to a constant, I prefer to have it dynamically picked up--so if the service is ever modified and the ports are changed, this will continue to work.

I don't see why this changes whether the sinks can be started in parallel.

Comment (vishh) [inline cluster/addons/cluster-monitoring/influxdb/heapster-controller.json]: The env variables won't be set for heapster pod unless the influxdb service has been created. Since we start all the monitoring services together, there is no guarantee that that influxdb service will be created before heapster.
Support for multi-port DNS is on the way. until then we can hardcode port 80. WDYT?

Comment (saad-ali) [inline cluster/addons/cluster-monitoring/influxdb/heapster-controller.json]: If that is the case, there would be no guarantee that Kube-DNS would resolve `monitoring-influxdb` correctly--i.e. if heapster pod gets created before the heapster service, Kube-DNS would not have a entry for `monitoring-influxdb`--we would end up with the same problem.

With the env variables approach, if the pod comes up before the service, and the variables are not set, heapster will terminate with error--and Kubernetes will create a new container which should have the variables set.

Comment (vishh) [inline cluster/addons/cluster-monitoring/influxdb/heapster-controller.json]: The heapster pod is supposed to handle backend failures. So if the dns name
does not resolve, it will retry internally before deciding to crash/fail.
Crashing without retires might confuse users.

Comment (saad-ali) [inline cluster/addons/cluster-monitoring/influxdb/heapster-controller.json]: That's reasonable. I'll go ahead and make the change.

**Comment by saad-ali** [comment]
PTAL.

Regarding JSON: There should be no reason that the cluster addons mechanism doesn't support both JSON and YAML. And it is good to have examples of both in our code. The change was suggested by @vishh -- his argument, as I understood it, was that YAML would internally be translated in to JSON anyway. That said, I don't feel too strongly about it and would be willing to change it back to YAML.

Rerunning E2E for verification.

**Comment by bgrant0607** [comment]
See #7257

**Comment by a-robinson** [comment]
I agree that the mechanism should support both, just wondering whether we should be writing any in JSON anymore. I'll defer to @bgrant0607 - do you care whether these cluster addon configs are in YAML vs JSON?

**Comment by saad-ali** [comment]
FYI, E2E is green:
```
Ran 38 of 43 Specs in 1687.865 seconds
SUCCESS! -- 38 Passed | 0 Failed | 1 Pending | 4 Skipped I0505 16:13:05.378063   24215 driver.go:96] All tests pass
```

**Comment by bgrant0607** [comment]
Let's please leave the configs in YAML.

**Comment by saad-ali** [comment]
PTAL Switched to YAML

E2E is clean:
```
Ran 38 of 43 Specs in 1653.461 seconds
SUCCESS! -- 38 Passed | 0 Failed | 1 Pending | 4 Skipped I0505 22:08:03.650021   24184 driver.go:96] All tests pass
```

File: cluster/addons/cluster-monitoring/google/heapster-controller.yaml [hunk 6]
```diff
@@ -0,0 +1,32 @@
+apiVersion: "v1beta3"
```
Comment (a-robinson) [inline cluster/addons/cluster-monitoring/google/heapster-controller.yaml]: You don't need to quote any of the strings in the yaml files (except perhaps if the string is "true", judging by the deleted files).

**Comment by a-robinson** [comment]
I'm sorry for asking you to change this again, but can you un-quote all the strings in the yaml files?

**Comment by saad-ali** [comment]
No worries. Quotes removed. PTAL

Re-confirmed E2E:
```
Ran 38 of 43 Specs in 1800.775 seconds
SUCCESS! -- 38 Passed | 0 Failed | 1 Pending | 4 Skipped I0506 14:54:35.184922   29922 driver.go:96] All tests pass
```

File: cluster/gce/config-test.sh [hunk 7]
```diff
@@ -49,8 +49,11 @@ ENABLE_DOCKER_REGISTRY_CACHE=true
 # Optional: Install node monitoring.
 ENABLE_NODE_MONITORING="${KUBE_ENABLE_NODE_MONITORING:-true}"
 
-# Optional: When set to true, heapster will be setup as part of the cluster bring up.
-ENABLE_CLUSTER_MONITORING="${KUBE_ENABLE_CLUSTER_MONITORING:-true}"
+# Optional: Cluster monitoring to setup as part of the cluster bring up:
+#   none     - No cluster monitoring setup 
+#   influxdb - Heapster, InfluxDB, and Grafana 
+#   google   - Heapster, Google Cloud Metrics, and Google Cloud Logging 
+ENABLE_CLUSTER_MONITORING="${KUBE_ENABLE_CLUSTER_MONITORING:-influxdb}"
```
Comment (vishh) [inline cluster/gce/config-test.sh]: Ideally we want to use `google` backend here as well. We need to update the monitoring e2e tests to not to depend on the default cluster addons before switching to google backend.

File: cluster/addons/cluster-monitoring/influxdb/grafana-service.yaml [hunk 8]
```diff
@@ -6,11 +6,9 @@ metadata:
     kubernetes.io/cluster-service: "true"
   name: monitoring-grafana
 spec: 
-  ports:
+  ports: 
     - port: 80
       targetPort: 8080
   selector: 
     name: influxGrafana
-    kubernetes.io/cluster-service: "true"
```
Comment (a-robinson) [inline cluster/addons/cluster-monitoring/influxdb/grafana-service.yaml]: Any particular reason for removing this?

File: cluster/gce/coreos/helper.sh [hunk 9]
```diff
@@ -33,7 +33,7 @@ SERVER_BINARY_TAR_URL: $(yaml-quote ${SERVER_BINARY_TAR_URL})
 SALT_TAR_URL: $(yaml-quote ${SALT_TAR_URL})
 PORTAL_NET: $(yaml-quote ${PORTAL_NET})
 ALLOCATE_NODE_CIDRS: $(yaml-quote ${ALLOCATE_NODE_CIDRS:-false})
-ENABLE_CLUSTER_MONITORING: $(yaml-quote ${ENABLE_CLUSTER_MONITORING:-false})
```
Comment (vishh) [inline cluster/gce/coreos/helper.sh]: Why is monitoring disabled for coreos? cc @dchen1107

File: cluster/addons/cluster-monitoring/influxdb/grafana-service.yaml (hunk 8 above)
Comment (a-robinson) [inline cluster/addons/cluster-monitoring/influxdb/grafana-service.yaml]: Ah, it looks like no selectors key off of this label. Nevermind

**Comment by vishh** [comment]
LGTM

File: cluster/gce/config-default.sh [hunk 10]
```diff
@@ -50,8 +50,11 @@ ENABLE_DOCKER_REGISTRY_CACHE=true
 # Optional: Install node monitoring.
 ENABLE_NODE_MONITORING="${KUBE_ENABLE_NODE_MONITORING:-true}"
 
-# Optional: When set to true, heapster, Influxdb and Grafana will be setup as part of the cluster bring up.
-ENABLE_CLUSTER_MONITORING="${KUBE_ENABLE_CLUSTER_MONITORING:-true}"
+# Optional: Cluster monitoring to setup as part of the cluster bring up:
+#   none     - No cluster monitoring setup 
+#   influxdb - Heapster, InfluxDB, and Grafana 
+#   google   - Heapster, Google Cloud Metrics, and Google Cloud Logging
```
Comment (a-robinson) [inline cluster/gce/config-default.sh]: Google Cloud Monitoring, not Google Cloud Metrics

File: cluster/gce/config-test.sh [hunk 11]
```diff
@@ -49,8 +49,11 @@ ENABLE_DOCKER_REGISTRY_CACHE=true
 # Optional: Install node monitoring.
 ENABLE_NODE_MONITORING="${KUBE_ENABLE_NODE_MONITORING:-true}"
 
-# Optional: When set to true, heapster will be setup as part of the cluster bring up.
-ENABLE_CLUSTER_MONITORING="${KUBE_ENABLE_CLUSTER_MONITORING:-true}"
+# Optional: Cluster monitoring to setup as part of the cluster bring up:
+#   none     - No cluster monitoring setup 
+#   influxdb - Heapster, InfluxDB, and Grafana 
+#   google   - Heapster, Google Cloud Metrics, and Google Cloud Logging
```
Comment (a-robinson) [inline cluster/gce/config-test.sh]: ditto

File: cluster/gce/config-test.sh (hunk 7 above)
Comment (saad-ali) [inline cluster/gce/config-test.sh]: Sure, we can do that separately.

File: cluster/gce/coreos/helper.sh (hunk 9 above)
Comment (saad-ali) [inline cluster/gce/coreos/helper.sh]: It's not. That's only the default value if `ENABLE_CLUSTER_MONITORING` is not set. The deployment process sets the value before this point in `cluster/gce/config-default.sh`

File: cluster/gce/config-default.sh (hunk 10 above)
Comment (saad-ali) [inline cluster/gce/config-default.sh]: Changed.

File: cluster/gce/config-test.sh (hunk 11 above)
Comment (saad-ali) [inline cluster/gce/config-test.sh]: Changed.

**Comment by a-robinson** [comment]
Thanks! LGTM.

**Comment by piosz** [comment]
cc @jszczepkowski
